File: ga/gpt3_5_result_2024_6/result/queue_gpt/48_queue.c
use std::ptr;

type ElementType = i32;

struct QueueRecord {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: Vec<ElementType>,
}

type Queue = Box<QueueRecord>;

const MIN_QUEUE_SIZE: usize = 5;

fn is_empty(Q: &Queue) -> bool {
    Q.size == 0
}

fn is_full(Q: &Queue) -> bool {
    Q.size == Q.capacity
}

fn create_queue(max_elements: usize) -> Queue {
    if max_elements < MIN_QUEUE_SIZE {
        panic!("Queue size is too small");
    }
    let buffer = vec![Default::default(); max_elements];
    Box::new(QueueRecord {
        capacity: max_elements,
        front: 1,
        rear: 0,
        size: 0,
        array: buffer,
    })
}

fn make_empty(Q: &mut Queue) {
    Q.size = 0;
    Q.front = 1;
    Q.rear = 0;
}

fn dispose_queue(Q: Queue) {
    drop(Q);
}

fn succ(value: usize, Q: &Queue) -> usize {
    if value + 1 == Q.capacity {
        return 0;
    } else {
        return value + 1;
    }
}

fn enqueue(X: ElementType, Q: &mut Queue) {
    if is_full(&Q) {
        panic!("Full queue");
    } else {
        Q.size += 1;
        Q.rear = succ(Q.rear, &Q);
        Q.array[Q.rear] = X;
    }
}

fn front(Q: &Queue) -> ElementType {
    if !is_empty(&Q) {
        return Q.array[Q.front];
    } else {
        panic!("Empty queue");
    }
}

fn dequeue(Q: &mut Queue) {
    if is_empty(&Q) {
        panic!("Empty queue");
    } else {
        Q.size -= 1;
        Q.front = succ(Q.front, &Q);
    }
}

fn front_and_dequeue(Q: &mut Queue) -> ElementType {
    let mut X: ElementType = Default::default();

    if is_empty(&Q) {
        panic!("Empty queue");
    } else {
        Q.size -= 1;
        X = Q.array[Q.front];
        Q.front = succ(Q.front, &Q);
    }
    return X;
}

fn main() {
    let mut Q = create_queue(12);

    for i in 0..10 {
        enqueue(i, &mut Q);
    }

    while !is_empty(&Q) {
        println!("{}", front(&Q));
        dequeue(&mut Q);
    }

    for i in 0..10 {
        enqueue(i, &mut Q);
    }

    while !is_empty(&Q) {
        println!("{}", front(&Q));
        dequeue(&mut Q);
    }

    dispose_queue(Q);
}

File: ga/4omini_result_2024_7/result/queue_mini/48_queue.rs
use std::alloc::{self, Layout};
use std::ptr;
use std::io::{self, Write};

type ElementType = i32;

#[derive(Debug)]
struct Queue {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: *mut ElementType,
}

impl Queue {
    const MIN_QUEUE_SIZE: usize = 5;

    fn new(max_elements: usize) -> Result<Self, &'static str> {
        if max_elements < Self::MIN_QUEUE_SIZE {
            return Err("Queue size is too small");
        }

        let layout = Layout::array::<ElementType>(max_elements).map_err(|_| "Out of space!!!")?;
        let array = unsafe { alloc::alloc(layout) };

        if array.is_null() {
            return Err("Out of space!!!");
        }

        Ok(Queue {
            capacity: max_elements,
            front: 0,
            rear: capacity - 1,
            size: 0,
            array,
        })
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    fn enqueue(&mut self, x: ElementType) -> Result<(), &'static str> {
        if self.is_full() {
            return Err("Full queue");
        }

        self.size += 1;
        self.rear = (self.rear + 1) % self.capacity;
        unsafe {
            *self.array.add(self.rear) = x;
        }

        Ok(())
    }

    fn front(&self) -> Result<ElementType, &'static str> {
        if self.is_empty() {
            return Err("Empty queue");
        }
        unsafe { Ok(*self.array.add(self.front)) }
    }

    fn dequeue(&mut self) -> Result<(), &'static str> {
        if self.is_empty() {
            return Err("Empty queue");
        }

        self.size -= 1;
        self.front = (self.front + 1) % self.capacity;

        Ok(())
    }

    fn front_and_dequeue(&mut self) -> Result<ElementType, &'static str> {
        if self.is_empty() {
            return Err("Empty queue");
        }

        let x = unsafe { *self.array.add(self.front) };
        self.size -= 1;
        self.front = (self.front + 1) % self.capacity;
        Ok(x)
    }
}

impl Drop for Queue {
    fn drop(&mut self) {
        unsafe {
            let layout = Layout::array::<ElementType>(self.capacity).unwrap();
            alloc::dealloc(self.array, layout);
        }
    }
}

fn main() {
    let mut queue = Queue::new(12).expect("Failed to create queue");

    // Enqueue numbers from 0 to 9
    for i in 0..10 {
        queue.enqueue(i).expect("Failed to enqueue");
    }

    // Dequeue and print numbers
    while !queue.is_empty() {
        let front_val = queue.front().expect("Failed to get front value");
        println!("{}", front_val);
        queue.dequeue().expect("Failed to dequeue");
    }

    // Enqueue numbers from 0 to 9 again
    for i in 0..10 {
        queue.enqueue(i).expect("Failed to enqueue");
    }

    // Dequeue and print numbers again
    while !queue.is_empty() {
        let front_val = queue.front().expect("Failed to get front value");
        println!("{}", front_val);
        queue.dequeue().expect("Failed to dequeue");
    }
}